### ISSUE TITLE
commitlog writes may block for a long time, timing us out

### ISSUE THREAD

**Issue by glommer** [scylladb/scylla#1691]
When we hit a certain limit of in-flight commitlog segment writes, we will stop accepting new requests. If that happens for a short period of time that's fine and desirable, and will condition the workload to be slower. If it happens for a long time, it can time us out and it means we should be probably doing something earlier about it.

Because the commitlog segments are not expected to be too big, it is counter intuitive that this is happening. It could be indicative of a bug somewhere, and/or situations where we need to wait for all segments to complete when one of them would be enough.

We need to investigate further to find out why and where the blocks happen and do something about it.

**Comment by glommer** [comment]
Managed to track this down to a _file.flush() issue.

So while the individual commitlog segments are written down very fast, flushing them can take a long time. As we may wait for those flushes to end before issuing new segments, writes can block for as long as the flushes block.

**Comment by glommer** [comment]
monitoring fdatasync() times with systemtap, we find out that they are finishing in dozens of milliseconds at the most. So that's not the issue. The problem seem to be that other iothread consumers are taking a long time. That is confirmed by systemtap, tracking suspicious calls like open() and rename(). Some of them are taking 10s of seconds to complete. As the syscall thread will serialize those calls, a flush that falls behind them will time us out.

Users that are suspicious that this could be the same problem they are seeing can use the attached systemtap script to verify if that's the case
[vfslat.stp.txt](https://github.com/scylladb/scylla/files/487562/vfslat.stp.txt)

**Comment by glommer** [comment]
Example trace obtained with the offtcputrace tool from bcc, when we spend more than 10s out of CPU:

```
    ffffffffb90c57b1 finish_task_switch
    ffffffffb97dffb5 schedule
    ffffffffb97e310c schedule_timeout
    ffffffffb97e1f12 __down
    ffffffffb90ea821 down
    ffffffffc046a9dc xfs_buf_lock
    ffffffffc046abfb _xfs_buf_find
    ffffffffc046ae4a xfs_buf_get_map
    ffffffffc046babd xfs_buf_read_map
    ffffffffc0499931 xfs_trans_read_buf_map
    ffffffffc044a561 xfs_da_read_buf
    ffffffffc0451390 xfs_dir3_leaf_read.constprop.16
    ffffffffc0452b90 xfs_dir2_leaf_lookup_int
    ffffffffc0452e0f xfs_dir2_leaf_lookup
    ffffffffc044d9d3 xfs_dir_lookup
    ffffffffc047d1d9 xfs_lookup
    ffffffffc0479e53 xfs_vn_lookup
    ffffffffb925347a path_openat
    ffffffffb9254a71 do_filp_open
    ffffffffb9242a94 do_sys_open
    ffffffffb9242b9e sys_open
    ffffffffb97e42b2 entry_SYSCALL_64_fastpath
    00007fb0698162ed [unknown]
    -                scylla (10327)
```

**Comment by glommer** [comment]
Running the load with blktrace, parsing with blkparse -st, we can see that the xfsaild thread can have a very high "Dispatch wait" -> the time a request takes between being queued and being dispatched. That's over 10 seconds.

The linux io scheduler in use is cfq. That thread is probably being starved by the CFQ as it prioritizes something else. Still not clear what.

Running the workload with the noop and deadline schedulers, the problem disappears.

**Comment by glommer** [comment]
Kernel bug.

Patch sent to Linux, and the maintainer has already queued it for commit. Will update this issue as soon as it lands somewhere.

**Comment by glommer** [comment]
The original reporter in scylla-users was using the deadline scheduler, so we also have another issue on top of the Linux Kernel one. This should stay open, and I will start investigation from scratch here.

**Comment by slivne** [comment]
@glommer any update ?

**Comment by glommer** [comment]
Big writes still timeout under certain circumstances but a lot has been
done.
My current work items are:
- we need to split the writes for big commitlog entries into smaller buffers
- we need to fix a bug with replay position reorders (#1802)
- we need to figure out what else is harming us

**Comment by glommer** [comment]
I consider this fixed now, after many patches.
The commitlog is still generating very large writes, but we can open a new issue for those.
